File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
from collections import deque
input = sys.stdin.readline

# n은 지도의 한 변의 길이
n = int(input())
map_list = []
for _ in range(n):
    map_list.append(list(map(int, input().strip())))

def bfs(x, y, visit):
    queue = deque()
    queue.append((x, y))
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    visit[x][y] = 1
    count = 0
    while queue:
        cur_x, cur_y = queue.popleft()
        count += 1
        for i in range(4):
            next_x = cur_x + dx[i]
            next_y = cur_y + dy[i]
            if 0 <= next_x < n and 0 <= next_y < n:
                if visit[next_x][next_y] == 0 and map_list[next_x][next_y] == 1:
                    visit[next_x][next_y] = 1
                    queue.append((next_x, next_y))
    return count

answer = []
visit = [[0] * n for _ in range(n)]
for i in range(n):
    for j in range(n):
        if map_list[i][j] == 1 and visit[i][j] == 0:
            answer.append(bfs(i, j, visit))
answer.sort()
print(len(answer))
for i in range(len(answer)):
    print(answer[i])

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;
class Pair {
    int x;
    int y;

    public Pair(int y, int x) {
        this.x = x;
        this.y = y;
    }
}

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        int[][] map = new int[n][n];
        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < n; j++) {
                map[i][j] = line.charAt(j) - '0';
            }
        }

        int[][] visited = new int[n][n];
        int[] dx = {1, 0, -1, 0};
        int[] dy = {0, 1, 0, -1};
        int group_cnt = 0;
        ArrayList<Integer> list = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (map[i][j] == 1 && visited[i][j] == 0) {
                    group_cnt++;
                    int house_cnt = 1;
                    ArrayDeque<Pair> queue = new ArrayDeque<>();
                    queue.offer(new Pair(i, j));
                    visited[i][j] = 1;
                    while (!queue.isEmpty()) {
                        Pair cur = queue.poll();
                        for (int k = 0; k < 4; k++) {
                            int next_x = cur.x + dx[k];
                            int next_y = cur.y + dy[k];
                            if (0 <= next_x && next_x < n && 0 <= next_y && next_y < n) {
                                if (map[next_y][next_x] == 1 && visited[next_y][next_x] == 0) {
                                    visited[next_y][next_x] = 1;
                                    house_cnt++;
                                    queue.offer(new Pair(next_y, next_x));
                                }
                            }
                        }
                    }
                    list.add(house_cnt);
                }
            }
        }

        Collections.sort(list);

        StringBuilder sb = new StringBuilder();
        sb.append(group_cnt).append("\n");
        for (int i = 0; i < list.size(); i++) {
            sb.append(list.get(i)).append("\n");
        }

        System.out.println(sb);
    }
}
